docs: added license faq (#10654)
hashicorp/nomad: website/content/docs/enterprise/license/faq.mdx

If you have built automation around `nomad license put`, this command will fail as it has been removed from Nomad v1.1.0+.

## Q: What are some other gotchas?

Operators at a company might not have access to the company's normal Nomad license. 
The operators may be relying on the now-removed 6-hour window in order to do daily work with Nomad. 
Running `nomad agent -dev` will fail for these users. In this case, they should get a trial license via the [trial license page](http://nomadproject.io/trial).

## Q: How can a user get their existing license to put on disk for an upgrade?

Use `nomad license get` to retrieve the license, then save the license to disk.

## Q: What happens when a license expires?
If a license expires after the server instance is joined to the cluster, the existing behavior is preserved: disallow Enterprise feature operations until a grace period expires, at which point the Nomad servers shut down. 
The grace period will be one day.

## Q: Will on-prem binaries with “baked-in” licenses work?

Yes. They will work as if they had a standard autoloaded license.

> 
> ## Q: Are license files cluster-specific or node-specific?
> 
> No. Licenses can be reused on different nodes and different clusters.

Reviewer: By saying "or" we make this Q accidentally either/or and that makes the answer ambiguous.

```suggestion

## Q: Are licenses specific to a cluster? Are licenses specific to a node?

No. The same license can be reused on multiple clusters and multiple nodes.
```

> ## Q: Where can users get a trial license for Nomad?
> Visit http://nomadproject.io/trial for a free 30-day trial license.

Reviewer: ```suggestion
## Q: Where can users get a trial license for Nomad?

Visit http://nomadproject.io/trial for a free 30-day trial license.
```

## Q: When a Nomad client agent restarts, will it still need a license file?

> Yes. Every time `nomad agent` runs as a server (or in -dev), it requires a license on disk.

Reviewer: ```suggestion
Yes. Every time `nomad agent` runs as a server (or in `-dev` mode), it requires a license on disk.
```

## Q: Will Nomad downgrades/rollbacks work?

When downgrading to a version of Nomad before 1.1.0, customers will need to follow the old  process for applying enterprise licenses to Nomad. 
This is detailed in the [0.12 portion of the upgrade guide](https://www.nomadproject.io/docs/upgrade/upgrade-specific#enterprise-licensing). Enterprise features will be enabled for 6 hours. 
During that time enterprise users should apply their license with the “nomad license put” command. 

> Licenses will not be autoloaded from the file system in earlier versions of Nomad.

Reviewer: Nomad cannot be downgraded at all. "Autoloaded" is not a user-facing concept.

Author: @tgross Can you elaborate on this a bit more?  The statement about Nomad not being down gradable is not making sense. Autoloaded is part of the boot-up sequence and outside the consumers control, but they are still available to roll back to a prior version of Nomad Enterprise.

Reviewer: There's two separate things I'm saying here:
* You cannot safely roll back to an earlier version of Nomad at all, licensing aside. It's not a supported operation. The Nomad state store has no guarantee of forward compatibility, so trying to roll back to an earlier version can result in a crash loop. And Nomad cannot be downgraded from ENT to OSS safely either ([example](https://github.com/hashicorp/nomad/issues/9774#issuecomment-772655733))
* "Autoloaded" is not a user-facing concept, by which I mean the term "autoloaded" is not in our documentation at all. That term is only used in the internal design documents. So it doesn't mean anything to tell a user that a license will not be "autoloaded". The language here should match the user-facing documentation.

Author: I see, so perhaps adding something along these lines. 

Q: Can I downgrade Nomad Enterprise from v1.1.0 to an earlier version?
Nomad cannot safely be rolled back to an earlier version. This is not a supported operation due to the Nomad state store not having a forward compatibility. The end result could potentially be a crash loop.

> ## Q: Is there a tutorial available for the license enablement?
> Please visit the [Enterprise License Tutorial](https://learn.hashicorp.com/tutorials/nomad/hashicorp-enterprise-license).

Reviewer: ```suggestion
## Q: Is there a tutorial available for the license enablement?

Please visit the [Enterprise License Tutorial](https://learn.hashicorp.com/tutorials/nomad/hashicorp-enterprise-license).
```

Is "enablement" the right word here?

---
layout: docs
page_title: Nomad Enterprise License FAQ
description: >-
  Frequently Asked Questions pertaining to Nomad Enterprise Licensing.
---

# Frequently Asked Questions (FAQ)

This FAQ is for the license changes introduced in Nomad Enterprise version v1.1.0+ent. 
Nomad Enterprise automatically load Nomad licenses when a Nomad server agent starts using Nomad Enterprise.

## Q: Can I get a quick summary of the Nomad changes?

Starting with Nomad Enterprise v1.1.0, the license enablement process is different.

HashiCorp Enterprise servers will no longer start without a license. When the server instance initializes the license watcher, the server reads from either an environment variable or file. If the license is missing, invalid, or expired, the server will immediately exit. 
This check is part of the server boot-up process.
In previous versions of HashiCorp enterprise products, one server could distribute a license to other servers via the Raft protocol. 
This will no longer work since each server must be able to find a valid license during the startup process.

> 

Reviewer: I don't have a strong opinion on the line spacing here but you might want to take a look at the preview at https://nomad-cij6qa2ft-hashicorp.vercel.app/docs/enterprise/license/faq#q-can-i-get-a-quick-summary-of-the-nomad-changes here because you need a double newline in order to break these into paragraphs.

## Q: What resources are available?

The list below is a great starting point for learning more about the license changes introduced in Nomad Enterprise v1.1.0

- [v1.1.0 Upgrade Guide](https://www.nomadproject.io/docs/upgrade/upgrade-specific#enterprise-licenses)

- [Nomad License Documentation](https://www.nomadproject.io/docs/enterprise/license)

- [License configuration values documentation](https://www.nomadproject.io/docs/configuration/server#license_path)

- [Install a HashiCorp Enterprise License Tutorial](https://learn.hashicorp.com/tutorials/nomad/hashicorp-enterprise-license)

## Q: Does this affect Nomad or older versions of Nomad?

This change only affects Nomad Enterprise version v1.1.0+ent and later.

## Q: Does the license have to be present on Nomad client nodes?

Only Nomad server nodes need to have the license present.

> Nomad agents running with the “-dev” flag will also need a license.

Reviewer: ```suggestion
Nomad agents running with the `-dev` flag will also need a license.
```

## Q: What are some pitfalls that could be experienced during an upgrade because of the license changes?

If you follow the steps in the [upgrade guide](https://www.nomadproject.io/docs/upgrade/upgrade-specific#enterprise-licenses), but have misconfigured the licensing steps, the server instances will stop during boot-up sequence. 

If you do not check cluster health after adding new servers, and verify that the new servers have joined properly, then you may accidentally take down old servers without replacing them.

If you are using a trial license in production, this could cause issues as the cluster will fail to start once the trial expires.

If you have built automation around `nomad license put`, this command will fail as it has been removed from Nomad v1.1.0+.

## Q: What are some other gotchas?

> You might not have access to your oragnization's Nomad license. 

Reviewer: Let's take a spellcheck pass:

```suggestion
You might not have access to your organization's Nomad license. 
```

The  6-hour temporary license window is now removed, and a license is required for proper boot-up.
Running `nomad agent -dev` without a license will fail.

## Q: How can a user get their existing license to put on disk for an upgrade?

> Use `nomad license get` to retrieve the license, then save the license to disk.

Reviewer: Unfortunately this isn't correct; the `nomad license get` returns the interpretation of the license (see https://www.nomadproject.io/docs/commands/license/get#examples) but not the license blob. A user will need to get their license from whereever they originally got it.

Author: @tgross I changed it to the following information.

## Q: How can a user get their existing license to put on disk for an upgrade?

If you are an existing HashiCorp enterprise customer you may contact your organization's customer success manager (CSM) or email support-softwaredelivery@hashicorp.com for information on how to get your organization's enterprise license.
You can use `nomad license get` to retrieve information about the license, but not the license itself.

Reviewer: That works 👍

## Q: What happens when a license expires?
If a license expires after a server node joins the cluster, the existing Enterprise features will continue to operate until a grace period expires. 
The grace period duration is 1 day, and upon the end of the grace period the Nomad servers will shut down. 

## Q: Will on-prem binaries with “baked-in” licenses work?

There is no impact to baked-in licensed binaries.

## Q: Are licenses cluster-specific? node-specific?

Licenses can be reused on different nodes and different clusters.

## Q: Where can users get a trial license for Nomad?

> Visit http://nomadproject.io/trial for a free 30-day trial license.

Reviewer: ```suggestion
Visit https://nomadproject.io/trial for a free 30-day trial license.
```
